components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java

```diff
@@ -735,20 +721,67 @@ public void deleteApplication(String appId) throws APIManagementException {
             WorkflowResponse response = removeApplicationWFExecutor.execute(workflow);
             workflow.setStatus(response.getWorkflowStatus());
 
-            if (WorkflowStatus.APPROVED == response.getWorkflowStatus()) {
+            if (WorkflowStatus.CREATED != response.getWorkflowStatus()) {
                 completeWorkflow(removeApplicationWFExecutor, workflow);
             } else {
                 //add entry to workflow table if it is only in pending state
                 addWorkflowEntries(workflow);
             }
-         
+            return response;
         } catch (APIMgtDAOException e) {
             String errorMsg = "Error occurred while deleting the application - " + appId;
             log.error(errorMsg, e);
             throw new APIManagementException(errorMsg, e, ExceptionCodes.APIMGT_DAO_EXCEPTION);
         }
     }
 
+    private void cleanupPendingTaskForApplicationDeletion(Application application) throws APIManagementException {
+        WorkflowExecutor createApplicationWFExecutor = WorkflowExecutorFactory.getInstance()
+                .getWorkflowExecutor(WorkflowConstants.WF_TYPE_AM_APPLICATION_CREATION);
+        WorkflowExecutor createSubscriptionWFExecutor = WorkflowExecutorFactory.getInstance()
+                .getWorkflowExecutor(WorkflowConstants.WF_TYPE_AM_SUBSCRIPTION_CREATION);
+        WorkflowExecutor updateApplicationWFExecutor = WorkflowExecutorFactory.getInstance()
+                .getWorkflowExecutor(WorkflowConstants.WF_TYPE_AM_APPLICATION_UPDATE);
+
+        String appId = application.getId();
+        // get subscriptions with pending status
+        List<Subscription> pendingSubscriptions = getApiSubscriptionDAO()
+                .getPendingAPISubscriptionsByApplication(appId);
+  
+        String applicationStatus = application.getStatus();
+
+        if (pendingSubscriptions == null || pendingSubscriptions.isEmpty()) {
+            // check whether application is on hold state
+            if (ApplicationStatus.APPLICATION_ONHOLD.equals(applicationStatus)) {
+                //delete pending tasks for application creation if any
+                cleanupPendingTask(createApplicationWFExecutor, appId,
+                        WorkflowConstants.WF_TYPE_AM_APPLICATION_CREATION);
+            }
+        } else {
+
+            // this means there are pending subscriptions. It also implies that there cannot be pending application
+            // approvals (cannot subscribe to a pending application)
+            for (Iterator iterator = pendingSubscriptions.iterator(); iterator.hasNext();) {
```

Comment:
My suggestion is to use while loop instead of for loop. :)

while(iterator.hasNext()){